samza-core/src/test/java/org/apache/samza/container/TestContainerHeartbeatMonitor.java

```diff
@@ -93,6 +103,31 @@ public void testDoesNotCallbackWhenHeartbeatAlive() throws InterruptedException
     verify(this.scheduler).shutdown();
   }
 
+  @Test
+  public void testReestablishConnectionWithNewAM() throws InterruptedException {
+    this.containerHeartbeatMonitor =
+        spy(new ContainerHeartbeatMonitor(this.onExpired, this.containerHeartbeatClient, this.scheduler, COORDINATOR_URL,
+            "0", coordinatorStreamStore, true, 10));
+    CoordinatorStreamValueSerde serde = new CoordinatorStreamValueSerde(SetConfig.TYPE);
+    byte[] newCoordinatorUrl = serde.toBytes("http://some-host-2.prod.linkedin.com");
+    ContainerHeartbeatResponse response1 = new ContainerHeartbeatResponse(false);
+    ContainerHeartbeatResponse response2 = new ContainerHeartbeatResponse(true);
+    when(this.containerHeartbeatClient.requestHeartbeat()).thenReturn(response1).thenReturn(response2);
+    when(this.containerHeartbeatMonitor.getContainerHeartbeatClient()).thenReturn(this.containerHeartbeatClient);
```

Comment:
This looks a bit strange as we are not recreating the heartbeat client for the sake of tests.
Ideally, you want to verify if a new client is created with a new URL and that gets used.

Reply:
the client's heartbeat response needs to be mocked and hence cant create a new one.
updated to return mock when using the new url.
Also added failure scenario tests